runtime/lua/vim/lsp/buf.lua

```diff
@@ -161,5 +161,21 @@ function M.clear_references()
   util.buf_clear_references()
 end
 
+function M.code_action(context)
+  validate { context = { context, 't', true } }
+  context = context or { diagnostics = util.get_line_diagnostics() }
+  local params = util.make_range_params()
```

Comment:
Would it possibly fix the problem that was mentioned about `virtualedit` if we added a `onemore` optional param to `make_range_params` that, if true, would check if you are at the end of the line and add one extra position if so? This would mean that you wouldn't have to actually go to the virtualedit location (or turn it on) but would instead get that by being at the end of the line?

I haven't tested or anything to see if that would be the case, just a thought.

Reply:
It is a bit more complicated than that unfortunately. We would have to detect whether there is a diagnostic one past the end of the line and only add one to the cursor position if there is. Also, what would happen if there is also a diagnostic at the last character in the line? Would we send two requests (one with the actual position and one one past) then? We would probably also have to change the underline for the diagnostic then, to keep the behaviors of diagnostics and code actions mutually consistent.

Summed up, keeping everything consistent that way would complicate the logic considerably and I am not sure if the result would be more/equally intuitive. I will take a look at what plugins do, as this may be a solved problem.

Reply:
Ah, I was suggesting that because I thought it had a start and end to request the codeActions. I did not read the spec yet, but I thought we could detect if we were on the end of the line, and if so, add `onemore` to the range so that it could include past the line without people actually having to enable that feature.

Does that make sense? I would be interested to see how other (particularly vim) plugins have handled this. Thanks for following up

Reply:
A code action is only returned if the range submitted to the server lies completely inside the range of the diagnostic. This ensures that if you make a selection and get a code action for that, it is valid for the entire selection. What that means for this specific case is:

Code:
```typescript
let a = 1
```

Diagnostic range:
```json
{
    "start" : {"line": 0, "character": 9},
    "end" : {"line": 0, "character": 9}
}
```
Proposed code action range (which gives no results):
```json
{
    "start" : {"line": 0, "character": 8},
    "end" : {"line": 0, "character": 9}
}
```
Range that does give results:
```json
{
    "start" : {"line": 0, "character": 9},
    "end" : {"line": 0, "character": 9}
}
```

Reply:
Ah, I see. I didn't realize that was the case. Thanks for explaining.